runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ExecutableStageTranslation.java

```diff
@@ -45,19 +53,60 @@ public static String generateNameFromStagePayload(ExecutableStagePayload stagePa
     RunnerApi.Components components = stagePayload.getComponents();
     final int transformsCount = stagePayload.getTransformsCount();
     sb.append("[").append(transformsCount).append("]");
-    sb.append("{");
+    Collection<String> names = new ArrayList<>();
     for (int i = 0; i < transformsCount; i++) {
       String name = components.getTransformsOrThrow(stagePayload.getTransforms(i)).getUniqueName();
-      // Python: Remove the 'ref_AppliedPTransform_' prefix which just makes the name longer
-      name = name.replaceFirst("^ref_AppliedPTransform_", "");
       // Java: Remove the 'ParMultiDo(Anonymous)' suffix which just makes the name longer
       name = name.replaceFirst("/ParMultiDo\\(Anonymous\\)$", "");
-      sb.append(name);
-      if (i + 1 < transformsCount) {
-        sb.append(", ");
-      }
+      names.add(name);
     }
-    sb.append("}");
+    sb.append(generateNameFromTransformNames(names, true));
     return sb.toString();
   }
+
+  public static String generateNameFromTransformNames(Collection<String> names, boolean truncate) {
+    Multimap<String, String> groupByOuter = LinkedHashMultimap.create();
+    for (String name : names) {
+      int index = name.indexOf('/');
+      if (index == -1) {
+        groupByOuter.put(name, "");
+      } else {
+        groupByOuter.put(name.substring(0, index), name.substring(index + 1));
+      }
+    }
+    if (groupByOuter.keySet().size() == 1) {
+      Map.Entry<String, Collection<String>> outer =
+          Iterables.getOnlyElement(groupByOuter.asMap().entrySet());
+      if (outer.getValue().size() == 1 && outer.getValue().contains("")) {
+        // Names consisted of a single name without any slashes.
+        return outer.getKey();
+      } else {
+        // Everything is in the same outer stage, enumerate at one level down.
+        return String.format(
+            "%s/%s", outer.getKey(), generateNameFromTransformNames(outer.getValue(), truncate));
```

Comment:
We go down a level here but only if we had multiple elements inside the same composite transform and no other composite transforms.